tools/expander.go

```diff
@@ -0,0 +1,207 @@
+	expanded := make([]any, 0, len(expandedList))
+	for _, x := range expandedList {
+		x := x
+		converted, err := GetMarshalledObject(x)
+		if err != nil {
+			return nil, err
+		}
+		expanded = append(expanded, converted)
+	}
+	result := make([]*I, 0, len(responseList))
+	for _, response := range responseList {
+		response := response
+		expandedMap, err := GetMappedJSONPaths(response, structWithPaths)
+		if err != nil {
+			return nil, err
+		}
+		expandedObjects := PopulateExpandedMap(expandedMap, expanded)
+		result = append(result, makeResult(response, expandedObjects))
+	}
+
+	return result, nil
+}
+
+// Convert the expanded response to the desired type.
+func GetConvertedExpanded[T any](e T, key string, getExpanded func(T) map[string]*any) *T {
+	var rv *T
+	if x, ok := getExpanded(e)[key]; ok {
+		if x == nil {
+			return nil
+		}
+		if x, ok := (*x).(*T); ok {
+			rv = x
+		}
+	}
+	return rv
+}
+
+// Populate the expanded map with references to the related objects.
+func PopulateExpandedMap(expandMap map[string]int, expanded []any) map[string]*any {
+	rv := make(map[string]*any)
+	for k, v := range expandMap {
+		rv[k] = &expanded[v]
+	}
+	return rv
+}
+func GetPaths[T any](v *T) []Path {
+	if v == nil {
+		return nil
+	}
+
+	var paths []Path
+	view := reflect.ValueOf(*v)
+	typeOfView := view.Type()
+
+	for i := 0; i < view.NumField(); i++ {
+		field := view.Field(i)
+		fieldName := typeOfView.Field(i).Name
+		// Check if the field name ends with "Path"
+		if strings.HasSuffix(fieldName, "Path") && !field.IsNil() {
+			pathValue := field.Elem().String()
+			paths = append(paths, Path{
+				Key:   fieldName,
+				Value: &pathValue,
+			})
+		}
+	}
+
+	return paths
+}
+
+/* Pass in the list you want to expand with a function to get the inner object that contains the expanded Paths
+ * For example:
+ * In the case of `v AppEntitlementWithUserBindings` you would pass in a getter that returns
+ * v.AppEntitlementView which is a `*AppEntitlementView`, expects a pointer
+ */
+func GetMappedJSONPaths[T, V any](item T, structWithPaths getStructWithPaths[T, V]) (map[string]int, error) {
+	fn := func(t T) []Path {
+		v := structWithPaths(t)
+		paths := GetPaths(v)
+		return paths
+	}
+	return mapJSONPaths[T](item, fn)
+}
+
+// Generic function to map the JSON paths to the respective index in the expanded list.
+func mapJSONPaths[T any](item T, getPaths func(T) []Path) (map[string]int, error) {
+	res := make(map[string]int)
+	for _, path := range getPaths(item) {
+		index, err := GetJSONPathIndex(path.Value)
+		if err != nil {
+			return nil, err
+		}
+		if index != -1 {
+			res[path.Key] = index
+		}
+	}
+
+	return res, nil
+}
+
+// Use reflection to call the `GetAtType` method on an pointer receiver, func GetAtType() *string.
+func GetAtTypeWithReflection[T any](input *T) *string {
```

Comment:
can't this be `T getAtType` with `type getAtType interface{ GetAtType() *string }`?

Reply:
The method set of *T does not automatically include pointer receiver methods declared on the T's concrete type A, and it does not implement interfaces that would be implemented by *A.
- https://stackoverflow.com/questions/71444847/go-with-generics-type-t-is-pointer-to-type-parameter-not-type-parameter

There is no real good way to constrain the struct and the pointer receiver on the struct, I am going to add a nil check to the func